akd/src/append_only_zks.rs

```diff
@@ -199,88 +206,108 @@ impl Azks {
         Ok(load_count)
     }
 
-    /// An azks is built both by the [crate::directory::Directory] and the auditor.
-    /// However, both constructions have very minor differences, and the append_only_usage
-    /// bool keeps track of this.
-    pub async fn batch_insert_leaves_helper<S: Database + Sync + Send>(
-        &mut self,
+    /// Inserts a batch of leaves recursively from a given node label.
+    #[async_recursion]
+    pub async fn recursive_batch_insert_leaves<S: Database + Sync + Send>(
         storage: &StorageManager<S>,
+        node_label: Option<NodeLabel>,
         insertion_set: Vec<Node>,
+        epoch: u64,
         append_only_exclude_usage: bool,
-    ) -> Result<(), AkdError> {
-        let (fallable_load_count, time_s) =
-            tic_toc(self.preload_nodes_for_insertion::<S>(storage, &insertion_set)).await;
-        let load_count = fallable_load_count?;
-
-        if let Some(time) = time_s {
-            info!(
-                "Preload of tree ({} objects loaded), took {} s",
-                load_count, time,
-            );
-        } else {
-            info!("Preload of tree ({} objects loaded) completed", load_count);
+    ) -> Result<(TreeNode, u64), AkdError> {
+        // Phase 1: Obtain the current root node of this subtree and count if a
+        // node is inserted.
+        let mut current_node;
+        let mut num_inserted;
+
+        match (node_label, &insertion_set[..]) {
+            (Some(node_label), _) => {
+                // Case 1: The node label is not None, meaning that there was an
+                // existing node at this level of the tree.
+                current_node =
+                    TreeNode::get_from_storage(storage, &NodeKey(node_label), epoch).await?;
+                num_inserted = 0;
```

Comment:
nit: just for readability, can we make this part of the else statement below? Then the match can be and assignment to num_inserted's initial value. Ex

```rust
let mut num_inserted = match (node_label, &insertion_set[..]) {
    (Some(...), ) => {

        if (blah) {
            1
        } else {
            0
        }
    }
    ...
}
```

Reply:
I've refactored this section and added an `else` block, hopefully it's clearer! I opted to not have the match expression return anything because the `current_node` variable would have to be assigned as well, and I think it's a bit neater to assign them together inside the match arms.